worker/upload.go

```diff
@@ -187,6 +214,166 @@ func (w *worker) initUploadManager(maxOverdrive uint64, overdriveTimeout time.Du
 	w.uploadManager = newUploadManager(w.bus, w, w, maxOverdrive, overdriveTimeout, logger)
 }
 
+func (w *worker) upload(ctx context.Context, r io.Reader, bucket, path string, opts ...UploadOption) (string, error) {
+	//  build upload config
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	// perform the upload
+	obj, partialSlabData, used, etag, err := w.uploadManager.Upload(ctx, r, cfg)
+	if err != nil {
+		return "", fmt.Errorf("couldn't upload object: %w", err)
+	}
+
+	// add parital slabs
+	if len(partialSlabData) > 0 {
+		partialSlabs, err := w.bus.AddPartialSlab(ctx, partialSlabData, uint8(cfg.rs.MinShards), uint8(cfg.rs.TotalShards), cfg.contractSet)
+		if err != nil {
+			return "", err
+		}
+		obj.PartialSlabs = partialSlabs
+	}
+
+	// persist the object
+	err = w.bus.AddObject(ctx, bucket, path, cfg.contractSet, obj, used)
+	if err != nil {
+		return "", fmt.Errorf("couldn't add object: %w", err)
+	}
+
+	// if packing was enabled try uploading packed slabs in a separate goroutine
+	if cfg.pack {
+		go w.tryUploadPackedSlabs(cfg.rs, cfg.contractSet)
+	}
+	return etag, nil
+}
+
+func (w *worker) uploadMultiPart(ctx context.Context, r io.Reader, bucket, path, uploadID string, partNumber int, opts ...UploadOption) (string, error) {
+	//  build upload config
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	// upload the part
+	obj, partialSlabData, used, etag, err := w.uploadManager.Upload(ctx, r, cfg)
+	if err != nil {
+		return "", fmt.Errorf("couldn't upload object: %w", err)
+	}
+
+	// add parital slabs
+	if len(partialSlabData) > 0 {
+		partialSlabs, err := w.bus.AddPartialSlab(ctx, partialSlabData, uint8(cfg.rs.MinShards), uint8(cfg.rs.TotalShards), cfg.contractSet)
+		if err != nil {
+			return "", err
+		}
+		obj.PartialSlabs = partialSlabs
+	}
+
+	// persist the part
+	err = w.bus.AddMultipartPart(ctx, bucket, path, cfg.contractSet, uploadID, partNumber, obj.Slabs, obj.PartialSlabs, etag, used)
+	if err != nil {
+		return "", fmt.Errorf("couldn't add multi part: %w", err)
+	}
+
+	// if packing was enabled try uploading packed slabs in a separate goroutine
+	if cfg.pack {
+		w.tryUploadPackedSlabs(cfg.rs, cfg.contractSet)
+	}
+	return etag, nil
+}
+
+func (w *worker) tryUploadPackedSlabs(rs api.RedundancySettings, contractSet string) {
+	key := fmt.Sprintf("%d-%d_%s", rs.MinShards, rs.TotalShards, contractSet)
+
+	w.uploadsMu.Lock()
+	if w.uploadingPackedSlabs[key] {
+		w.uploadsMu.Unlock()
+		return
+	}
+	w.uploadingPackedSlabs[key] = true
+	w.uploadsMu.Unlock()
+
+	// make sure we mark uploading packed slabs as false when we're done
+	defer func() {
+		w.uploadsMu.Lock()
+		w.uploadingPackedSlabs[key] = false
+		w.uploadsMu.Unlock()
+	}()
+
+	// keep uploading packed slabs until we're done
+	packedSlabsFound := true
+	for packedSlabsFound {
```

Comment:
This looks like the calling thread is going to upload packed slabs until there are no more? Why wouldn't you keep the current behaviour of every thread fetching up to 2 packed slabs and uploading them after the regular upload?

This way the load isn't spread out quite as much and a single call might block for a while.

Reply:
The idea is to upload packed slabs in a goroutine, thus not blocking the caller, until there are no more packed slabs left. In one of the two methods I added (I think it was `w.multipartUpload`) I was missing to `go w.tryUploadPackedSlabs` because I toggled it on/off to see if tests would pass and then forgot to re-add. Aside from that oversight it's better not to block the original caller right?